File: coding/src/com/algos/core/li10_tree/bt/revised/Bt_L124_BinaryTreeMaximumPathSum.java
package  com.algos.core.li10_tree.bt.revised;

import  com.algos.core.li30_model.TreeNode;

public class Bt_L124_BinaryTreeMaximumPathSum {
    public static void main(String[] args) {
        Bt_L124_BinaryTreeMaximumPathSum obj = new Bt_L124_BinaryTreeMaximumPathSum();

        TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

        System.out.println(obj.maxPathSum(root));
    }

    // Problem: https://leetcode.com/problems/binary-tree-maximum-path-sum/
    // Idea: Idea is similar to finding max diameter. 
    // In paths, we can ignore the negative sum paths by taking max with 0 to skip negative path sums.
    public int maxPathSum(TreeNode root) {
        int[] maxPathSum = new int[]{Integer.MIN_VALUE};
        maxPathSum(root, maxPathSum);
        return maxPathSum[0];
    }
    
    private int maxPathSum(TreeNode root, int[] maxPathSum){
        if(root == null)
            return 0;
        
        int leftSum = Math.max(0, maxPathSum(root.left, maxPathSum));
        int rightSum = Math.max(0, maxPathSum(root.right, maxPathSum));
        int pathSum = root.val+leftSum+rightSum;
        maxPathSum[0] = Math.max(maxPathSum[0], pathSum);

        return root.val+Math.max(leftSum, rightSum);
    }
}

File: dsa/patterns/10_trees/bt/python/bt_l124__binary_tree_maximum_path_sum.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/binary-tree-maximum-path-sum/
# Idea: Similar to finding max diameter.
# Ignore negative path sums by taking max with 0 to skip negative paths.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def max_path_sum(root):
    max_sum = float('-inf')

    def helper(node):
        nonlocal max_sum
        if not node:
            return 0

        left_sum = max(0, helper(node.left))
        right_sum = max(0, helper(node.right))

        # Max path sum through current node
        path_sum = node.val + left_sum + right_sum
        max_sum = max(max_sum, path_sum)

        # Return max sum path extending to parent
        return node.val + max(left_sum, right_sum)

    helper(root)
    return max_sum


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(max_path_sum(root))  # Expected: 21 (path: 2 -> 3 -> 6 -> 5 -> 4)
